src/frontend/hooks/useLocation.ts

```diff
@@ -0,0 +1,113 @@
+import {useFocusEffect} from '@react-navigation/native';
+import CheapRuler from 'cheap-ruler';
+import {
+  watchPositionAsync,
+  hasServicesEnabledAsync,
+  useForegroundPermissions,
+  type LocationObject,
+  Accuracy,
+} from 'expo-location';
+import React from 'react';
+
+interface LocationOptions {
+  /** Only update location if it has changed by at least this distance in meters (or maxTimeInterval has passed) */
+  maxDistanceInterval: number;
+  /** If the location has updated by less than this distance in meters, don't update even if maxTimeInterval has passed */
+  minDistanceInterval?: number;
+  /** Minimum time to wait between each update in milliseconds. */
+  minTimeInterval?: number;
+  /** Maximum time to wait between each update in milliseconds - updates could take longer than this, but if a new location is available we will always get it after this time */
+  maxTimeInterval?: number;
+}
+
+const MAX_RULER_LAT_DISPLACEMENT = 1;
+
+// Timeout between location updates --> means location was probably turned off
+// so we need to check it.
+const LOCATION_TIMEOUT = 10000;
+
+export function useLocation(options: LocationOptions) {
+  const {minTimeInterval: timeInterval = 200, ...debounceOptions} = options;
+  const [location, setLocation] = React.useState<LocationObject | undefined>();
+
+  const [permissions] = useForegroundPermissions();
+
+  useFocusEffect(
+    React.useCallback(() => {
+      if (!permissions || !permissions.granted) return;
+
+      let ignore = false;
+      const LocationSubscriptionProm = watchPositionAsync(
+        {accuracy: Accuracy.BestForNavigation},
```

Comment:
not sure if this is helpful or any clearer, but could set `distanceInterval: debounceOptions.minDistanceInterval`, which then saves part of a check in the debounce function (can remove the `distance > minDistanceInterval` part)

Reply:
no, unfortunately we cannot do that as that would mean the `debounce` function would only get called when the user has moved 15m. We are also updating the location if the user has moved more than 1m after 2 seconds. That wouldn't be called if we add `distanceInterval`

Reply:
sorry I meant the 1 meter distance with my original comment, if that wasn't clear. but yeah given your reasoning, can disregard

Reply:
wait, but them I think your right, we could do that...

Reply:
ah ok, un-resolved this in case you want to consider it then 😄 my head keeps getting mixed up about the requirements. it may make sense, but also not the biggest deal left alone

Reply:
addressed with [this PR](https://github.com/digidem/CoMapeo-mobile/pull/107/commits/fa942647c4ca4cb1be9dfbf9cc3a53aed71d531a), But also adjusted the code, see [this comment](https://github.com/digidem/CoMapeo-mobile/pull/107/files#r1414621394) for reasoning